File: ui/src/optimization-final.ts
/** Closed, row-free final-holdout boundary. No scores or native payloads. */
export interface FinalIdentity { id: string; fingerprint: string }
export interface AgentFinalScope {
  run: FinalIdentity; launch: FinalIdentity; executionHead: string; adaptiveClosedAt: string;
  completion: FinalIdentity; iteration: FinalIdentity; trainingBindingFingerprint: string;
  developmentResult: FinalIdentity; comparisonBaselineRevision: FinalIdentity; benchmark: FinalIdentity;
  scientificProject: FinalIdentity; protocol: FinalIdentity; candidate: FinalIdentity; model: FinalIdentity;
  dataset: FinalIdentity & { datasetId: string; projectId: string; number: number };
  finalSuite: string; baselineFinalReport: FinalIdentity; metricContractFingerprint: string;
  maximumEvaluationSeconds: number; fingerprint: string;
}
export interface AgentFinalAuthorization { id: string; scope: AgentFinalScope; authorizedBy: string; createdAt: string; fingerprint: string }
export interface AgentFinalDispatch { authorization: FinalIdentity; reportId: string; assessmentId: string; createdAt: string; fingerprint: string }
export interface AgentFinalReceipt {
  dispatch: FinalIdentity; report: FinalIdentity; assessment: FinalIdentity; accepted: boolean;
  reportCreatedAt: string; createdAt: string; fingerprint: string;
}
export interface AgentFinalView {
  state: "authorized" | "outcome_unknown" | "completed";
  execution: { authorization: AgentFinalAuthorization; dispatch: AgentFinalDispatch | null; result: AgentFinalReceipt | null };
}
export interface AgentFinalReview { token: string; scope: AgentFinalScope }
export interface AgentFinalPromotionRequest { baselineRevisionId: string; receiptFingerprint: string }

export function finalObject(value: unknown, keys: string[]): Record<string, unknown> {
  if (!value || typeof value !== "object" || Array.isArray(value) || Object.keys(value).length !== keys.length
    || Object.keys(value).some(key => !keys.includes(key))) throw new Error("Invalid final-acceptance record.");
  return value as Record<string, unknown>;
}
export function finalUuid(value: unknown): string {
  if (typeof value !== "string" || !/^[0-9a-f]{8}-[0-9a-f]{4}-[1-8][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i.test(value)) throw new Error("Invalid final-acceptance identity.");
  return value.toLowerCase();
}
export function finalFingerprint(value: unknown): string {
  if (typeof value !== "string" || !/^sha256:[a-f0-9]{64}$/.test(value)) throw new Error("Invalid final-acceptance fingerprint.");
  return value;
}
function text(value: unknown): string {
  if (typeof value !== "string" || !value.trim() || value !== value.trim() || value.length > 240 || /[\u0000-\u001f]/.test(value)) throw new Error("Invalid final-acceptance text.");
  return value;
}
function timestamp(value: unknown): string { const result = text(value); if (!Number.isFinite(Date.parse(result))) throw new Error("Invalid final-acceptance time."); return result; }
function count(value: unknown): number { if (typeof value !== "number" || !Number.isSafeInteger(value) || value < 1) throw new Error("Invalid final-acceptance limit."); return value; }
function identity(value: unknown): FinalIdentity { const item = finalObject(value, ["id", "fingerprint"]); return { id: finalUuid(item.id), fingerprint: finalFingerprint(item.fingerprint) }; }
function matches(a: FinalIdentity, b: FinalIdentity): boolean { return a.id === b.id && a.fingerprint === b.fingerprint; }

export function parseAgentFinalScope(value: unknown, projectId: string, runId: string): AgentFinalScope {
  const item = finalObject(value, ["run", "launch", "executionHead", "adaptiveClosedAt", "completion", "iteration", "trainingBindingFingerprint", "developmentResult", "comparisonBaselineRevision", "benchmark", "scientificProject", "protocol", "candidate", "model", "dataset", "finalSuite", "baselineFinalReport", "metricContractFingerprint", "maximumEvaluationSeconds", "fingerprint"]);
  const dataset = finalObject(item.dataset, ["id", "datasetId", "projectId", "number", "fingerprint"]);
  const result: AgentFinalScope = {
    run: identity(item.run), launch: identity(item.launch), executionHead: finalFingerprint(item.executionHead), adaptiveClosedAt: timestamp(item.adaptiveClosedAt),
    completion: identity(item.completion), iteration: identity(item.iteration), trainingBindingFingerprint: finalFingerprint(item.trainingBindingFingerprint),
    developmentResult: identity(item.developmentResult), comparisonBaselineRevision: identity(item.comparisonBaselineRevision), benchmark: identity(item.benchmark),
    scientificProject: identity(item.scientificProject), protocol: identity(item.protocol), candidate: identity(item.candidate), model: identity(item.model),
    dataset: { id: finalUuid(dataset.id), datasetId: finalUuid(dataset.datasetId), projectId: finalUuid(dataset.projectId), number: count(dataset.number), fingerprint: finalFingerprint(dataset.fingerprint) },
    finalSuite: text(item.finalSuite), baselineFinalReport: identity(item.baselineFinalReport), metricContractFingerprint: finalFingerprint(item.metricContractFingerprint),
    maximumEvaluationSeconds: count(item.maximumEvaluationSeconds), fingerprint: finalFingerprint(item.fingerprint),
  };
  if (result.run.id !== runId || result.dataset.projectId !== projectId) throw new Error("Final acceptance belongs to another project or run.");
  return result;
}
export function parseAgentFinalAuthorization(value: unknown, projectId: string, runId: string): AgentFinalAuthorization {
  const item = finalObject(value, ["id", "scope", "authorizedBy", "createdAt", "fingerprint"]);
  const result = { id: finalUuid(item.id), scope: parseAgentFinalScope(item.scope, projectId, runId), authorizedBy: text(item.authorizedBy), createdAt: timestamp(item.createdAt), fingerprint: finalFingerprint(item.fingerprint) };
  if (Date.parse(result.createdAt) < Date.parse(result.scope.adaptiveClosedAt)) throw new Error("Final consent precedes adaptive completion.");
  return result;
}
export function parseAgentFinalView(value: unknown, projectId: string, runId: string): AgentFinalView | null {
  if (value === null) return null;
  const item = finalObject(value, ["state", "execution"]), execution = finalObject(item.execution, ["authorization", "dispatch", "result"]);
  const authorization = parseAgentFinalAuthorization(execution.authorization, projectId, runId);
  let dispatch: AgentFinalDispatch | null = null, result: AgentFinalReceipt | null = null;
  if (execution.dispatch !== null) {
    const raw = finalObject(execution.dispatch, ["authorization", "reportId", "assessmentId", "createdAt", "fingerprint"]);
    dispatch = { authorization: identity(raw.authorization), reportId: finalUuid(raw.reportId), assessmentId: finalUuid(raw.assessmentId), createdAt: timestamp(raw.createdAt), fingerprint: finalFingerprint(raw.fingerprint) };
    if (!matches(dispatch.authorization, authorization) || dispatch.reportId === dispatch.assessmentId || Date.parse(dispatch.createdAt) < Date.parse(authorization.createdAt)) throw new Error("Final dispatch differs from consent.");
  }
  if (execution.result !== null) {
    const raw = finalObject(execution.result, ["dispatch", "report", "assessment", "accepted", "reportCreatedAt", "createdAt", "fingerprint"]);
    if (typeof raw.accepted !== "boolean") throw new Error("Invalid final-acceptance decision.");
    result = { dispatch: identity(raw.dispatch), report: identity(raw.report), assessment: identity(raw.assessment), accepted: raw.accepted, reportCreatedAt: timestamp(raw.reportCreatedAt), createdAt: timestamp(raw.createdAt), fingerprint: finalFingerprint(raw.fingerprint) };
    if (!dispatch || !matches(result.dispatch, { id: authorization.id, fingerprint: dispatch.fingerprint }) || result.report.id !== dispatch.reportId || result.assessment.id !== dispatch.assessmentId
      || Date.parse(result.reportCreatedAt) < Date.parse(dispatch.createdAt) || Date.parse(result.createdAt) < Date.parse(result.reportCreatedAt)) throw new Error("Final result differs from its reserved evidence.");
  }
  const state = result ? "completed" : dispatch ? "outcome_unknown" : "authorized";
  if (item.state !== state) throw new Error("Final-acceptance state differs from persisted custody.");
  return { state, execution: { authorization, dispatch, result } };
}

File: crates/project-workspace-core/src/optimization_final.rs
//! Separate, row-free final-holdout authority after adaptive work has closed.
//! This never changes the zero-sealed-allowance iteration protocol.
use crate::{
    BoundIdentity, FinalEvaluationAuthorization, Invalid, OptimizationLaunchAuthorization,
    ProjectOptimizationRunState, ProjectOptimizationRunView,
    optimization_execution::AgentExecutionState,
    optimization_iteration::ProjectOptimizationIteration,
    optimization_iteration_execution::{IterationDevelopmentResult, IterationTrainingBinding},
    optimization_loop::IterationCompletion,
    require,
};
use chrono::{DateTime, Utc};
use dataset_core::versions::DatasetVersionRef;
use encoder_experiment_core::{
    domain::ExternalProjectSnapshot, journal::ExperimentEvent, protocol::ExperimentProtocol,
};
use serde::{Deserialize, Serialize};
use uuid::Uuid;

pub struct SelectedFinalEvidence<'a> {
    pub run: &'a ProjectOptimizationRunView,
    pub launch: &'a OptimizationLaunchAuthorization,
    /// The latest completion, deeply revalidated from all original journals.
    pub completion: &'a IterationCompletion,
    pub iteration: &'a ProjectOptimizationIteration,
    pub training: &'a IterationTrainingBinding,
    pub project: &'a ExternalProjectSnapshot,
    pub protocol: &'a ExperimentProtocol,
    pub events: &'a [ExperimentEvent],
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AgentFinalScope {
    pub run: BoundIdentity,
    pub launch: BoundIdentity,
    pub execution_head: String,
    pub adaptive_closed_at: DateTime<Utc>,
    pub completion: BoundIdentity,
    pub iteration: BoundIdentity,
    pub training_binding_fingerprint: String,
    pub development_result: BoundIdentity,
    pub comparison_baseline_revision: BoundIdentity,
    pub benchmark: BoundIdentity,
    pub scientific_project: BoundIdentity,
    pub protocol: BoundIdentity,
    pub candidate: BoundIdentity,
    pub model: BoundIdentity,
    pub dataset: DatasetVersionRef,
    pub final_suite: String,
    /// Identity only. Sealed baseline scores never enter the handoff.
    pub baseline_final_report: BoundIdentity,
    pub metric_contract_fingerprint: String,
    pub maximum_evaluation_seconds: u64,
    pub fingerprint: String,
}

impl AgentFinalScope {
    pub fn bind(evidence: SelectedFinalEvidence<'_>) -> Result<Self, Invalid> {
        let SelectedFinalEvidence {
            run,
            launch,
            completion,
            iteration,
            training,
            project,
            protocol,
            events,
        } = evidence;
        run.run.validate(launch)?;
        completion.validate_identity()?;
        training.validate_for(iteration, launch)?;
        let execution = run
            .agent_execution
            .as_ref()
            .ok_or_else(|| Invalid("Agent execution is missing".into()))?;
        let settings = launch
            .scope
            .agentic
            .as_ref()
            .ok_or_else(|| Invalid("Agent settings are missing".into()))?;
        require(
            settings.permits_final_evaluation()
                && launch.scope.final_evaluation
                    == FinalEvaluationAuthorization::SelectedCandidateOnce
                && launch.scope.limits.maximum_final_evaluations == 1,
            "Diagnostic runs cannot authorize final holdout",
        )?;
        require(
            match (run.state, execution.state) {
                (ProjectOptimizationRunState::AgentCompleted, AgentExecutionState::Completed) => {
                    completion.end.is_some()
                        && execution.completion.as_ref() == Some(&completion.identity())
                }
                (
                    ProjectOptimizationRunState::AgentBudgetExhausted,
                    AgentExecutionState::BudgetExhausted,
                ) => execution.completion.is_none(),
                _ => false,
            },
            "Finish adaptive work before authorizing final holdout",
        )?;
        let selected = completion
            .selected
            .as_ref()
            .ok_or_else(|| Invalid("No development-eligible candidate was selected".into()))?;
        let result = IterationDevelopmentResult::from_journal(training, project, protocol, events)?;
        let preparation = run
            .preparation
            .as_ref()
            .ok_or_else(|| Invalid("Prepared inputs are missing".into()))?;
        require(
            completion.run == run.run.identity()
                && iteration.run == completion.run
                && selected.iteration == bound(iteration.id, &iteration.fingerprint)
                && selected.candidate == training.candidate
                && selected.dataset == training.qualified_dataset
                && selected.result == bound(result.experiment_run_id, &result.fingerprint)
                && training.training_dataset == training.qualified_dataset
                && result.development_passed
                && completion.created_at <= execution.updated_at
                && iteration.benchmark == preparation.benchmark
                && protocol.sealed_suite_key == preparation.final_suite,
            "Final holdout must use the exact development-selected full-data candidate",
        )?;
        let mut value = Self {
            run: run.run.identity(),
            launch: bound(launch.id, &launch.fingerprint),
            execution_head: execution.head_fingerprint.clone(),
            completion: completion.identity(),
            adaptive_closed_at: execution.updated_at,
            iteration: selected.iteration.clone(),
            training_binding_fingerprint: training.fingerprint.clone(),
            development_result: selected.result.clone(),
            comparison_baseline_revision: iteration.comparison_baseline_revision.clone(),
            benchmark: iteration.benchmark.clone(),
            scientific_project: training.scientific_project.clone(),
            protocol: training.protocol.clone(),
            candidate: training.candidate.clone(),
            model: bound(result.output.model.id, &result.output.model.fingerprint),
            dataset: selected.dataset.clone(),
            final_suite: protocol.sealed_suite_key.clone(),
            baseline_final_report: bound(
                protocol.baseline_sealed_report.id,
                &protocol.baseline_sealed_report.fingerprint,
            ),
            metric_contract_fingerprint: protocol.metric_contract.fingerprint.clone(),
            maximum_evaluation_seconds: protocol.maximum_evaluation_seconds,
            fingerprint: String::new(),
        };
        value.fingerprint = value.reproduce()?;
        value.validate_identity()?;
        Ok(value)
    }

    pub fn reproduce(&self) -> Result<String, Invalid> {
        reproduce(self)
    }
    pub fn validate_identity(&self) -> Result<(), Invalid> {
        for identity in [
            &self.run,
            &self.launch,
            &self.completion,
            &self.iteration,
            &self.development_result,
            &self.comparison_baseline_revision,
            &self.benchmark,
            &self.scientific_project,
            &self.protocol,
            &self.candidate,
            &self.model,
            &self.baseline_final_report,
        ] {
            identity.validate("Final holdout identity")?;
            require(
                Uuid::parse_str(&identity.id).is_ok_and(|id| !id.is_nil()),
                "Final holdout requires exact UUIDs",
            )?;
        }
        for hash in [
            &self.execution_head,
            &self.training_binding_fingerprint,
            &self.metric_contract_fingerprint,
        ] {
            crate::validate_hash(hash)?;
        }
        self.dataset.validate().map_err(invalid)?;
        crate::validate_name(&self.final_suite)?;
        require(
            self.maximum_evaluation_seconds > 0 && self.fingerprint == self.reproduce()?,
            "Final holdout scope changed",
        )
    }
}

#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AgentFinalAuthorization {
    pub id: Uuid,
    pub scope: AgentFinalScope,
    pub authorized_by: String,
    pub created_at: DateTime<Utc>,
    pub fingerprint: String,
}

impl AgentFinalAuthorization {
    pub fn create(
        id: Uuid,
        scope: AgentFinalScope,
        authorized_by: String,
        created_at: DateTime<Utc>,
    ) -> Result<Self, Invalid> {
        let mut value = Self {
            id,
            scope,
            authorized_by,
            created_at,
            fingerprint: String::new(),
        };
        value.fingerprint = reproduce(&value)?;
        value.validate_identity()?;
        Ok(value)
    }
    pub fn validate_identity(&self) -> Result<(), Invalid> {
        self.scope.validate_identity()?;
        crate::validate_name(&self.authorized_by)?;
        require(
            !self.id.is_nil()
                && self.created_at >= self.scope.adaptive_closed_at
                && self.fingerprint == reproduce(self)?,
            "Final authorization changed or predates adaptive completion",
        )
    }
}

fn reproduce(value: &impl Serialize) -> Result<String, Invalid> {
    let mut value = serde_json::to_value(value).map_err(invalid)?;
    value.as_object_mut().expect("record").remove("fingerprint");
    artifact_core::fingerprint(&value).map_err(invalid)
}
fn bound(id: Uuid, fingerprint: &str) -> BoundIdentity {
    BoundIdentity {
        id: id.to_string(),
        fingerprint: fingerprint.into(),
    }
}
fn invalid(error: impl std::fmt::Display) -> Invalid {
    Invalid(error.to_string())
}
